internal/core/src/storage/Event.cpp

```diff
@@ -0,0 +1,357 @@
+    memcpy(res.data(), &start_timestamp, sizeof(start_timestamp));
+    offset += sizeof(start_timestamp);
+    memcpy(res.data(), &end_timestamp, sizeof(end_timestamp));
+    offset += sizeof(end_timestamp);
+    memcpy(res.data() + offset, payload_buffer.data(), payload_buffer.size());
+
+    return res;
+}
+
+InsertEvent::InsertEvent(PayloadInputStream* input, DataType data_type) {
+    event_header = EventHeader(input);
+    auto event_data_length = event_header.event_length_ - event_header.next_position_;
+    event_data = InsertEventData(input, event_data_length, data_type);
+}
+
+std::vector<uint8_t>
+InsertEvent::Serialize() {
+    auto data = event_data.Serialize();
+    int data_size = data.size();
+
+    event_header.event_type_ = EventType::InsertEvent;
+    event_header.next_position_ = GetEventHeaderSize(event_header);
+    event_header.event_length_ = event_header.next_position_ + data_size;
+    auto header = event_header.Serialize();
+    int header_size = header.size();
+
+    int len = header_size + data_size;
+    std::vector<uint8_t> res(len);
+    int offset = 0;
+    memcpy(res.data() + offset, header.data(), header_size);
+    offset += header_size;
+    memcpy(res.data() + offset, data.data(), data_size);
+
+    return res;
+}
+
+DescriptorEvent::DescriptorEvent(PayloadInputStream* input) {
+    event_header = EventHeader(input);
+    event_data = DescriptorEventData(input);
+}
+
+std::vector<uint8_t>
+DescriptorEvent::Serialize() {
+    auto data = event_data.Serialize();
+    int data_size = data.size();
+
+    event_header.event_type_ = EventType::DescriptorEvent;
+    event_header.next_position_ = GetEventHeaderSize(event_header);
+    event_header.event_length_ = event_header.next_position_ + data_size;
+    auto header = event_header.Serialize();
+    int header_size = header.size();
+
+    int len = header_size + data_size + sizeof(MAGIC_NUM);
+    std::vector<uint8_t> res(len);
+    int offset = 0;
+    memcpy(res.data(), &MAGIC_NUM, sizeof(MAGIC_NUM));
+    offset += sizeof(MAGIC_NUM);
+    memcpy(res.data() + offset, header.data(), header_size);
+    offset += header_size;
+    memcpy(res.data() + offset, data.data(), data_size);
+
+    return res;
+}
+
+LocalInsertEvent::LocalInsertEvent(PayloadInputStream* input, DataType data_type) {
+    auto ret = input->Read(sizeof(row_num), &row_num);
+    AssertInfo(ret.ok(), "read input stream failed");
+    ret = input->Read(sizeof(dimension), &dimension);
+    AssertInfo(ret.ok(), "read input stream failed");
+    int data_size = milvus::datatype_sizeof(data_type) * row_num;
+    auto insert_data_bytes = input->Read(data_size);
+    auto insert_data = reinterpret_cast<const uint8_t*>(insert_data_bytes.ValueOrDie()->data());
+    std::shared_ptr<arrow::ArrayBuilder> builder = nullptr;
+    if (milvus::datatype_is_vector(data_type)) {
+        builder = CreateArrowBuilder(data_type, dimension);
+    } else {
+        builder = CreateArrowBuilder(data_type);
+    }
```

Comment:
extract it as a util function maybe better 🤔

Reply:
> extract it as a util function maybe better thinking

？